core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,378 @@
+        invalidPropNames.foreach(props.remove)
+        throw new ConfigException(s"$errorMessage: $invalidPropNames")
+      }
+    }
+    checkInvalidProps(nonDynamicConfigs(props), "Cannot update these configs dynamically")
+    checkInvalidProps(securityConfigsWithoutListenerPrefix(props),
+      "Security configs can be dynamically updated only per-listener using the listener prefix")
+    validateConfigTypes(props)
+    val newProps = mutable.Map[String, String]()
+    newProps ++= staticBrokerConfigs
+    if (perBrokerConfig) {
+      overrideProps(newProps, dynamicDefaultConfigs)
+      overrideProps(newProps, props.asScala)
+    } else {
+      checkInvalidProps(perBrokerConfigs(props),
+        "Cannot update these configs at default cluster level, broker id must be specified")
+      overrideProps(newProps, props.asScala)
+      overrideProps(newProps, dynamicBrokerConfigs)
+    }
+    processConfig(newProps, validateOnly = true)
+  }
+
+  private def perBrokerConfigs(props: Properties): Set[String] = {
+    val configNames = props.asScala.keySet
+    configNames.intersect(PerBrokerConfigs) ++ configNames.filter(ListenerConfigRegex.findFirstIn(_).nonEmpty)
+  }
+
+  private def nonDynamicConfigs(props: Properties): Set[String] = {
+    props.asScala.keySet.intersect(DynamicConfig.Broker.nonDynamicProps)
+  }
+
+  private def securityConfigsWithoutListenerPrefix(props: Properties): Set[String] = {
+    DynamicSecurityConfigs.filter(props.containsKey)
+  }
+
+  private def validateConfigTypes(props: Properties): Unit = {
+    val baseProps = new Properties
+    props.asScala.foreach {
+      case (ListenerConfigRegex(baseName), v) => baseProps.put(baseName, v)
+      case (k, v) => baseProps.put(k, v)
+    }
+    DynamicConfig.Broker.validate(baseProps)
+  }
+
+  private def removeInvalidConfigs(props: Properties, perBrokerConfig: Boolean): Unit = {
+    try {
+      validateConfigTypes(props)
+      props.asScala
+    } catch {
+      case e: Exception =>
+        val invalidProps = props.asScala.filter { case (k, v) =>
+          val props1 = new Properties
+          props1.put(k, v)
+          try {
+            validateConfigTypes(props1)
+            false
+          } catch {
+            case _: Exception => true
+          }
+        }
+        invalidProps.foreach(props.remove)
+        val configSource = if (perBrokerConfig) "broker" else "default cluster"
+        error(s"Dynamic $configSource config contains invalid values: $invalidProps, these configs will be ignored", e)
+    }
+  }
+
+  /**
+    * Converts a map of config (key, value) pairs to a map of strings where each value
+    * is converted to a string. This method should be used with care since it stores
+    * actual password values to String. Values from this map should never be used in log entries.
+    */
+  private def configToMap(config: Map[String, _]): immutable.Map[String, String] = {
+    config.filter(_._2 != null).map {
```

Comment:
I still think it would be reasonable to move the conversion logic over to `ConfigDef`